File: Snooper/src/main/java/com/prateekj/snooper/infra/CurrentActivityManager.java
package com.prateekj.snooper.infra;


import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import java.util.ArrayList;
import java.util.List;

public class CurrentActivityManager implements Application.ActivityLifecycleCallbacks {

  private static CurrentActivityManager sInstance;

  public interface Listener {
    void currentActivity(Activity activity);
  }

  private final List<Listener> listeners = new ArrayList<>();

  private CurrentActivityManager(Application application) {
    application.registerActivityLifecycleCallbacks(this);
  }

  public static CurrentActivityManager getInstance(Application application) {
    if (sInstance == null) {
      sInstance = new CurrentActivityManager(application);
    }
    return sInstance;
  }


  public void registerListener(Listener listener) {
    listeners.add(listener);
  }

  public void unregisterListener(Listener listener) {
    listeners.remove(listener);
  }


  @Override
  public void onActivityResumed(Activity activity) {
    notifyListeners(activity);
  }


  @Override
  public void onActivityPaused(Activity activity) {
    notifyListeners(activity);
  }

  private void notifyListeners(Activity activity) {
    for (Listener listener : listeners) {
      listener.currentActivity(activity);
    }
  }


  @Override
  public void onActivityStopped(Activity activity) {
  }

  @Override
  public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
  }

  @Override
  public void onActivityStarted(Activity activity) {
  }

  @Override
  public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
  }

  @Override
  public void onActivityDestroyed(Activity activity) {
  }
}

File: Snooper/src/main/java/com/prateekj/snooper/infra/CurrentActivityManager.kt
package com.prateekj.snooper.infra


import android.app.Activity
import android.app.Application
import android.app.Application.ActivityLifecycleCallbacks
import android.os.Bundle

import java.util.ArrayList

class CurrentActivityManager private constructor(application: Application) :
  ActivityLifecycleCallbacks {

  private val listeners = ArrayList<Listener>()

  interface Listener {
    fun currentActivity(activity: Activity?)
  }

  init {
    application.registerActivityLifecycleCallbacks(this)
  }

  fun registerListener(listener: Listener) {
    listeners.add(listener)
  }

  fun unregisterListener(listener: Listener) {
    listeners.remove(listener)
  }


  override fun onActivityResumed(activity: Activity?) {
    notifyListeners(activity)
  }


  override fun onActivityPaused(activity: Activity?) {
    notifyListeners(activity)
  }

  private fun notifyListeners(activity: Activity?) {
    for (listener in listeners) {
      listener.currentActivity(activity)
    }
  }

  override fun onActivityStopped(activity: Activity?) {}

  override fun onActivityCreated(activity: Activity?, savedInstanceState: Bundle?) {}

  override fun onActivityStarted(activity: Activity?) {}

  override fun onActivitySaveInstanceState(activity: Activity?, outState: Bundle?) {}

  override fun onActivityDestroyed(activity: Activity?) {}

  companion object {

    private var INSTANCE: CurrentActivityManager? = null

    fun getInstance(application: Application): CurrentActivityManager {
      return INSTANCE ?: CurrentActivityManager(application).also { INSTANCE = it }
    }
  }
}
